tests/xbot/common/logic/LatchTest.java

```diff
@@ -0,0 +1,76 @@
+package xbot.common.logic;
+
+import org.junit.Test;
+
+import java.util.Observable;
+import java.util.Observer;
+
+import static junit.framework.TestCase.assertTrue;
+
+public class LatchTest {
+
+    @Test
+    public void testRisingEdge() {
+        Latch latch = new Latch(false, Latch.EdgeType.RisingEdge);
+        MockLatch mockLatch = new MockLatch();
+        latch.addObserver(mockLatch);
+        latch.setValue(true);
+        assertTrue(mockLatch.getEdgeType() == Latch.EdgeType.RisingEdge);
+    }
+
+    @Test
+    public void testFallingEdge() {
+        Latch latch = new Latch(true, Latch.EdgeType.FallingEdge);
+        MockLatch mockLatch = new MockLatch();
+        latch.addObserver(mockLatch);
+        latch.setValue(false);
+        assertTrue(mockLatch.getEdgeType() == Latch.EdgeType.FallingEdge);
+    }
+
+    @Test
+    public void testRiseFallRiseObservingRisingEdge() {
+        Latch latch = new Latch(false, Latch.EdgeType.RisingEdge);
+        MockLatch mockLatch = new MockLatch();
+        latch.addObserver(mockLatch);
+        assertTrue(mockLatch.getEdgeType() == null);
+        latch.setValue(true);
+        assertTrue(mockLatch.getEdgeType() == Latch.EdgeType.RisingEdge);
+        latch.setValue(false);
+        assertTrue(mockLatch.getEdgeType() == Latch.EdgeType.RisingEdge);
+        latch.setValue(true);
+        assertTrue(mockLatch.getNumTimesUpdated() == 2);
+    }
+
+    @Test
+    public void testFallRiseFallObservingFallingEdge() {
+        Latch latch = new Latch(true, Latch.EdgeType.FallingEdge);
+        MockLatch mockLatch = new MockLatch();
+        latch.addObserver(mockLatch);
+        latch.setValue(false);
+        assertTrue(mockLatch.getEdgeType() == Latch.EdgeType.FallingEdge);
+        latch.setValue(true);
+        assertTrue(mockLatch.getEdgeType() == Latch.EdgeType.FallingEdge);
+        latch.setValue(false);
+        assertTrue(mockLatch.getNumTimesUpdated() == 2);
+    }
+}
+
+class MockLatch implements Observer {
```

Comment:
I think this class should be renamed `LatchTestObserver` or similar. It would be good if `getEdgeType` were called `getLastUpdateEdgeType` to make it clear that it is reporting the state from the Latch.